### PR TITLE
feat: sort assets by market cap

### LINKED ISSUE #288: Add market cap data to redux store
Now we have caips on asset and market and market cap data add the market cap data to the store, then we can join it up and use it in the UI

### PR DESCRIPTION
**Pull request by 0xdef1cafe**
## Description

sorts asset list by market cap

note: this only fetches the market cap for the first 250 assets to avoid rate limiting. we can fetch as many as we want with parallel requests by changing one line. assets without market cap are shown in the same list sorted by name and symbol. no one is going to scroll past the 250th asset to find things manually, they're just going to search.

## Notice

Before submitting a pull request, please make sure you have answered the following:

- [ ] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?
- [ ] Do all new and existing tests pass? Does the linter pass?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #288 

## Testing

Please outline all testing steps

1. load assets page
2. ethereum should flip bitcoin

## Screenshots (if applicable)

Screen Shot 2021-11-08 at 2 15 09 PM

### CODE REVIEW

**Review by elmutt — APPROVED** [review]

File: src/components/AssetSearch/selectors/selectAndSortAssets/selectAndSortAssets.ts [hunk 1]
```diff
@@ -1,18 +1,37 @@
...
+    if (marketCap) {
+      // we only fetch market data for the top 250 assets
+      // and want this to be fairly performant so do some mutatey things
+      const assetsByCAIP19 = assetsEntries.reduce<Record<string, FullAsset>>((acc, [, cur]) => {
+        acc[cur.caip19] = cur
+        return acc
+      }, {})
+      const caip19ByMarketCap = Object.keys(marketCap)
+      const sortedWithMarketCap = caip19ByMarketCap.reduce<FullAsset[]>((acc, cur) => {
+        acc.push(assetsByCAIP19[cur])
```
Comment (natven) [inline src/components/AssetSearch/selectors/selectAndSortAssets/selectAndSortAssets.ts:22]: any potential for assetsByCAIP19 to return null/undefined?

File: src/components/AssetSearch/selectors/selectAndSortAssets/selectAndSortAssets.ts [hunk 2]
```diff
@@ -1,18 +1,37 @@
...
+        delete assetsByCAIP19[cur]
+        return acc
+      }, [])
+      const remainingSortedNoMarketCap = sortBy(Object.values(assetsByCAIP19), ['name', 'symbol'])
```
Comment (natven) [inline src/components/AssetSearch/selectors/selectAndSortAssets/selectAndSortAssets.ts:26]: i'd probably default to symbol alone but no biggie

File: src/pages/Dashboard/hooks/useTotalBalance/useTotalBalance.ts [hunk 3]
```diff
@@ -15,7 +15,7 @@ export const useTotalBalance = (accounts: ReturnType<typeof flattenTokenBalances
     Object.values(accounts).forEach(account => {
       if (!account) return
       const identifier = account.contract ?? account.chain
-      const accountMarketData = marketData[identifier]
+      const accountMarketData = marketData.marketData[identifier]
```
Comment (natven) [inline src/pages/Dashboard/hooks/useTotalBalance/useTotalBalance.ts:18]:
```suggestion
      const accountMarketData = marketData?.marketData?.[identifier]
```
maybe?

**Review by natven — APPROVED** [review]

File: src/components/AssetSearch/selectors/selectAndSortAssets/selectAndSortAssets.ts [hunk 4]
```diff
@@ -1,18 +1,37 @@
+import { CoinGeckoMarketCapResult } from '@shapeshiftoss/types'
```
[the remark this answers is not in the document]
Comment (0xdef1cafe) [reply src/components/AssetSearch/selectors/selectAndSortAssets/selectAndSortAssets.ts:1]: yeah, i didn't like this leaking when i implemented it either. the next change we make to the market service we'll refactor it like the asset services with adapters and data sources and hide this behind a common type that's standard enough that we can pull the same fields from multiple sources.

File: src/pages/Dashboard/hooks/useTotalBalance/useTotalBalance.ts (hunk 3 above)
Comment (0xdef1cafe) [reply src/pages/Dashboard/hooks/useTotalBalance/useTotalBalance.ts:18]: nah not this one, the inner object is always defined and the string lookup will return undefined which gets caught below

File: src/components/AssetSearch/selectors/selectAndSortAssets/selectAndSortAssets.ts (hunk 2 above)
Comment (0xdef1cafe) [reply src/components/AssetSearch/selectors/selectAndSortAssets/selectAndSortAssets.ts:26]: ya was just copypastaing existing logic

File: src/components/AssetSearch/selectors/selectAndSortAssets/selectAndSortAssets.ts (hunk 1 above)
Comment (0xdef1cafe) [reply src/components/AssetSearch/selectors/selectAndSortAssets/selectAndSortAssets.ts:22]: highly unlikely coingecko doesn't have asset data for the top 250 by market cap, but i'll add a check for this if some coin pricing is temporarily unavailable

Comment (0xdef1cafe) [reply src/components/AssetSearch/selectors/selectAndSortAssets/selectAndSortAssets.ts]: done

**Pull request merged**